File: nfs-server/modules/IMU/Mag/GY80_HMC5883L.py
#!/usr/bin/python
"""
*************************************************
* @Project: Self Balance  
* @Platform: Raspberry PI 2 B+                       
* @Description: GY80 API - Orientation sensor via I2C bus
*               HMC5883L (Magnetometer/Compass sensor)
* @Owner: Guilherme Chinellato                   
* @Email: guilhermechinellato@gmail.com                               
*************************************************
"""

from Utils.i2c.Adafruit_I2C import Adafruit_I2C
from IMU.Accel.GY80_ADXL345 import GY80_Accelerometer_ADXL345
from IMU.constants import *
import math
import time
import os
import sys
from Utils.traces.trace import *

class GY80_Magnetometer_HMC5883L():
    def __init__(self, gauss=1.3, busnum=-1, cal=False, debug=False):
        self.i2c = Adafruit_I2C(HMC5883L_ADDRESS, busnum, debug)
        self.debug = debug
        self.gauss = gauss
        self.cal = cal

        self.magOffset = None
        self.scale = None
        self.declination = None
        self.Heading = None

        #Set to Data output rate @ 75Hz
        self.i2c.write8(HMC5883L_ConfigurationRegisterA, 0x18)
        #Set 1.3 gain LSb / Gauss 1090 (default)
        self.setScale(gauss)        
        #Continuous sampling
        self.i2c.write8(HMC5883L_ModeRegister, HMC5883L_MeasurementContinuous)

        #Hard calibration offset    
        if cal == True:
            self.magOffset = self.getOffset()
            raw_input("Press any key to continue...")
        else:
            #read offset from a file
            self.magOffset = self.restoreOffsetCalibration()

        logging.info("Magnetometer module initialized")

    def setScale(self, gauss):
        if gauss == 0.88:
			self.scale = 0.73
        elif gauss == 1.3:
			scale_reg = 0x01
			self.scale = 0.92
        elif gauss == 1.9:
			scale_reg = 0x02
			self.scale = 1.22
        elif gauss == 2.5:
			scale_reg = 0x03
			self.scale = 1.52
        elif gauss == 4.0:
			scale_reg = 0x04
			self.scale = 2.27
        elif gauss == 4.7:
			scale_reg = 0x05
			self.scale = 2.56
        elif gauss == 5.6:
			scale_reg = 0x06
			self.scale = 3.03
        elif gauss == 8.1:
			scale_reg = 0x07
			self.scale = 4.35
		
        scale_reg = scale_reg << 5
        self.i2c.write8(HMC5883L_ConfigurationRegisterB , scale_reg)
	
    #Read raw data
    def readRawMag(self, scaled=False):
        bytes = self.i2c.readList(HMC5883L_AxisXDataRegisterMSB, 6)

        mag = []
        #read x,y,z raw data
        for i in range(0, 6, 2):
            #merge MSB|LSB
            g = (bytes[i] << 8) | bytes[i+1]
            #negative values
            if g > 32767:
                g -= 65536
            mag.append(g)

        #According datasheet, read sequence is X, Z and Y
        rawMag = [0,0,0]
        rawMag[X] = mag[0]
        rawMag[Y] = mag[2]
        rawMag[Z] = mag[1]

        if (self.debug):
            logging.debug(("Mag Raw: x:%d, y:%d, z:%d" % (rawMag[X], rawMag[Y], rawMag[Z])))

        return rawMag

    #Normalize data: v'= v/\v\ - unit vector
    def normalizeRawData(self, rawMag):
        normMag = [0,0,0]
        normMag[X] = rawMag[X]/math.sqrt(rawMag[X]*rawMag[X] + rawMag[Y]*rawMag[Y] + rawMag[Z]*rawMag[Z])
        normMag[Y] = rawMag[Y]/math.sqrt(rawMag[X]*rawMag[X] + rawMag[Y]*rawMag[Y] + rawMag[Z]*rawMag[Z])
        normMag[Z] = rawMag[Z]/math.sqrt(rawMag[X]*rawMag[X] + rawMag[Y]*rawMag[Y] + rawMag[Z]*rawMag[Z])

        if (self.debug):
            logging.debug(("Mag Normalized: x:%.3f, y:%.3f, z:%.3f" % (normMag[X], normMag[Y], normMag[Z])))

        return normMag

    #Convert raw mag to Gauss
    def scaleMag(self, rawMag):        
        #In the event the ADC reading overflows or underflows for the given channel, or if there is a math overflow during the bias
        #measurement, this data register will contain the value -4096.
        scaledMag = [0,0,0]
        if (rawMag[X] == -4096):
		    scaledMag[X] = None
        else:
		    scaledMag[X] = round(rawMag[X] * self.scale, 4)
		
        if (rawMag[Y] == -4096):
		    scaledMag[Y] = None
        else:
		    scaledMag[Y] = round(rawMag[Y] * self.scale, 4)
		
        if (rawMag[Z] == -4096):
		    scaledMag[Z] = None
        else:
		    scaledMag[Z] = round(rawMag[Z] * self.scale, 4)

        if (self.debug):
            logging.debug(("Mag Scaled: x:%.3f, y:%.3f, z:%.3f (mG)" % (scaledMag[X], scaledMag[Y], scaledMag[Z])))

        return scaledMag
	
    # Set actual declination according localization
    def setDeclination(self, degree, minutes):
		self.declination = (degree+minutes/60) * (math.pi/180)

    # Offset calibration - Get max and min values from calibration 
    def getOffset(self):
        magXmax = 0
        magYmax = 0
        magZmax = 0
        magXmin = 0
        magYmin = 0
        magZmin = 0

        logging.info("Mag calibration")
        raw_input("Rotate the device through the 3-axes. Press any key to start...")
        
        for i in range(300):
            rawMag = self.readRawMag()
            logging.info(("Calibrating Mag...%d: " % (i)))
            logging.info(("MAG X MAX: %0.3f MIN %0.3f " % (magXmax, magXmin)))
            logging.info(("MAG Y MAX: %0.3f MIN %0.3f " % (magYmax, magYmin)))
            logging.info(("MAG Z MAX: %0.3f MIN %0.3f " % (magZmax, magZmin)))

            if (rawMag[X] > magXmax):
                magXmax = rawMag[X]
            if (rawMag[Y] > magYmax):
                magYmax = rawMag[Y]
            if (rawMag[Z] > magZmax):
                magZmax = rawMag[Z]
            if (rawMag[X] < magXmin):
                magXmin = rawMag[X]
            if (rawMag[Y] < magYmin):
                magYmin = rawMag[Y]
            if (rawMag[Z] < magZmin):
                magZmin = rawMag[Z]

            time.sleep(0.1)  

        magOffset = [0,0,0]
        magOffset[X] = (magXmin + magXmax)/2
        magOffset[Y] = (magYmin + magYmax)/2 
        magOffset[Z] = (magZmin + magZmax)/2 

        if (self.debug):
            logging.debug(("Mag Offset: x:%0.3f, y:%0.3f, z:%0.3f" % (magOffset[X], magOffset[Y], magOffset[Z])))

        #Store offset in a file
        self.storeOffsetCalibration(magOffset)

        return magOffset  

    def storeOffsetCalibration(self, magOffset): 
        msg = str(magOffset).strip('[]')       
        with open(magOffsetPath, 'w') as offsetFile:
            offsetFile.write(msg)

        logging.info("Mag offset value written properly!")
        logging.info(("Mag Offset WRITE: %s" % (msg)))

    def restoreOffsetCalibration(self):        
        if os.path.isfile(magOffsetPath):
            with open(magOffsetPath, 'r') as offsetFile:
                msg = offsetFile.read()

            msg = msg.split(",") 
            magOffset = []           
            for item in msg:
                magOffset.append(float(item))

            logging.info("Mag offset value read properly!")
            logging.info(("Mag Offset READ: x:%0.3f, y:%0.3f, z:%0.3f" % (magOffset[X], magOffset[Y], magOffset[Z])))
        else:
            logging.warning((magOffsetPath + str(" does not exist, please calibrate again!")))
            sys.exit()

        return magOffset

    def hardCalibration(self, rawMag, magOffset):
        calMag = [0,0,0]
        calMag[X] = rawMag[X] - magOffset[X]
        calMag[Y] = rawMag[Y] - magOffset[Y]
        calMag[Z] = rawMag[Z] - magOffset[Z]

        if (self.debug):
            logging.debug(("Mag Calibrated: x:%0.3f, y:%0.3f, z:%0.3f" % (calMag[X], calMag[Y], calMag[Z])))

        return calMag 

    def tiltCompensation(self, accelRollRad, accelPitchRad, rawMag):
        cosPitch = math.cos(accelPitchRad) 
        sinPitch = math.sin(accelPitchRad)
        cosRoll = math.cos(accelRollRad)
        sinRoll = math.sin(accelRollRad)

        # Tilt compensation
        compMag = [0,0,0]
        compMag[X] = (rawMag[X]*cosPitch) + (rawMag[Z]*sinPitch)
        compMag[Y] = (rawMag[X]*sinRoll*sinPitch) + (rawMag[Y]*cosRoll) - (rawMag[Z]*sinRoll*cosPitch)
        compMag[Z] = (-rawMag[X]*cosRoll*sinPitch) + (rawMag[Y]*sinRoll) - (rawMag[Z]*cosRoll*cosPitch)

        if (self.debug):
            logging.debug(("Mag Tilt: x:%0.2f, y:%0.2f" % (compMag[X], compMag[Y])))

        return compMag  

    def getHeading(self, compMagX, compMagY, decliFlag=True):
        #Only needed if the heading value does not increase when the magnetometer is rotated clockwise
        #compMagY = -compMagY;

        headingRad = math.atan2(compMagY, compMagX)

        if decliFlag == True:
            headingRad += self.declination

		# Correct for reversed heading
        if(headingRad < 0):
            headingRad += 2*math.pi
			
		# Check for wrap and compensate
        if(headingRad > 2*math.pi):
            headingRad -= 2*math.pi
	
		# Convert to degrees from radians
        headingDeg = headingRad * RAD_TO_DEG

        # Convert to degrees amd minutes
        degrees = math.floor(headingDeg)
        minutes = round(((headingDeg - degrees) * 60))

        if (self.debug):
            logging.debug(("Mag Heading: x:%0.2f" % (headingDeg)))

        self.Heading = headingDeg
        return headingDeg
     
def TestHMC5883L():
    # Enable Magnetometer and Accelerometer
    mag = GY80_Magnetometer_HMC5883L(debug=True, cal=False)
    accel = GY80_Accelerometer_ADXL345()

    LP = 0.2

    while True:
        #
        # Accelerometer
        # 
        scaledAcc = accel.scaleAcc(accel.readRawAcc())
        normAcc = accel.normalizeRawData(scaledAcc) 
        accel.calculateAngles(normAcc)

        #
        # Magnetometer
        # 

        # Magnetic declination 
        # Santo Andre @ Brazil:
        # Declination: -21 7 (Negative WEST)
        mag.setDeclination(-21,7)
        
        # Read mag value
        rawMag = mag.readRawMag()
        calMag = mag.hardCalibration(rawMag, mag.magOffset)
        scaledMag = mag.scaleMag(calMag)
        normMag = mag.normalizeRawData(scaledMag)

        print "Accel: Roll: " + str(round(accel.Roll*RAD_TO_DEG,2)) + ", Pitch: " + str(round(accel.Pitch*RAD_TO_DEG,2))
        compMag = mag.tiltCompensation(accel.Roll, accel.Pitch, normMag)   
        # Tilted        
        yaw1 = mag.getHeading(compMag[X], compMag[Y]) 
        # Not-Tilted        
        yaw2 = mag.getHeading(normMag[X], normMag[Y]) 

        time.sleep(LP)

if __name__ == '__main__':
    TestHMC5883L()
 

File: Arduino/src/IMU/Mag/GY80_HMC5883L.cpp
/**
*************************************************
* @Project: Self Balance
* @Platform: Arduino Nano ATmega328
* @Description: Magnetometer sensor
* @Owner: Guilherme Chinellato
* @Email: guilhermechinellato@gmail.com
*************************************************
 */

#include <Arduino.h>
#include <Wire.h>
#include <EEPROM.h>
#include "GY80_HMC5883L.h"

/*
 * object constructor:
 * initialize I2C communication
 * write 0x00 to register 0x01 (place the part in continuous measurement mode)
 * write 0x70 to register 0x00 (Configuration Register A - default configuration)
 * write appropriate value depending on HMC5883L_DefaultFieldRange to register 0x01(configuration register B)
 */
HMC5883L::HMC5883L()
{
	Wire.begin();

	writeTo(HMC5883L_ModeRegister, HMC5883L_MeasurementContinuous);
	writeTo(HMC5883L_ConfigurationRegisterA, 0x18);

	if (HMC5883L_DefaultFieldRange==0.88) // Nominal gain configuration (HMC5883L_ConfigurationRegisterB)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 1370.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x00);
	}
	else if (HMC5883L_DefaultFieldRange==1.3)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 1090.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x20);
	}
	else if (HMC5883L_DefaultFieldRange==1.9)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 820.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x40);
	}
	else if (HMC5883L_DefaultFieldRange==2.5)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 660.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x60);
	}
	else if (HMC5883L_DefaultFieldRange==4.0)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 440.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x80);
	}
	else if (HMC5883L_DefaultFieldRange==4.7)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 390.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xA0);
	}
	else if (HMC5883L_DefaultFieldRange==5.6)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 330.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xC0);
	}
	else if (HMC5883L_DefaultFieldRange==8.1)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 230.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xE0);
	}
	else // out of range - return to defaults // default configuration: field range 1.3Ga
	{
		HMC5883L_SCALE_FACTOR = (1000 / 1090);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x20);
	}
}

/*
 * writeTo method:
 * write 'val' in 'address' register (using I2C bus <Wire> lib)
 */
void HMC5883L::writeTo(byte address, byte val)
{
	Wire.beginTransmission(HMC5883L_DEVICE);
	Wire.write(address);
	Wire.write(val);
	Wire.endTransmission();
}

/*
 * readFrom method:
 * read 'num' bytes from starting 'address' register onwards and stores in '_buff[]'
 * (using I2C bus <Wire> lib)
 */
void HMC5883L::readFrom(byte address, int num, byte _buff[])
{
	Wire.beginTransmission(HMC5883L_DEVICE);
	Wire.write(address);
	Wire.endTransmission();

	Wire.beginTransmission(HMC5883L_DEVICE);
	Wire.requestFrom(HMC5883L_DEVICE, num);

	int i = 0;
	while(Wire.available())
	{
		_buff[i] = Wire.read();
		i++;
	}
	/*if(i != num){
		status = HMC5883L_ERROR;
		error_code = HMC5883L_READ_ERROR;
	}*/
	Wire.endTransmission();
}

/*
 * setGain method:
 * set HMC5883L_SCALE_FACTOR based on 'fieldRange' value
 * write to HMC5883L_ConfigurationRegisterB register the appropriate value for the specified 'fieldRange'
 */
void HMC5883L::setGain(float fieldRange)
{
	if (fieldRange==0.88) // Nominal gain configuration (HMC5883L_ConfigurationRegisterB)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 1370.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x00);
	}
	else if (fieldRange==1.3)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 1090.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x20);
	}
	else if (fieldRange==1.9)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 820.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x40);
	}
	else if (fieldRange==2.5)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 660.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x60);
	}
	else if (fieldRange==4.0)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 440.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x80);
	}
	else if (fieldRange==4.7)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 390.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xA0);
	}
	else if (fieldRange==5.6)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 330.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xC0);
	}
	else if (fieldRange==8.1)
	{
		HMC5883L_SCALE_FACTOR = (1000.0f / 230.0f);
		writeTo(HMC5883L_ConfigurationRegisterB, 0xE0);
	}
	else // out of range - return to defaults // default configuration: field range 1.3Ga
	{
		HMC5883L_SCALE_FACTOR = (1000 / 1090);
		writeTo(HMC5883L_ConfigurationRegisterB, 0x20);
	}
}

/*
 * read method:
 * store sensor data(x,y,z) in 'orientationVector'
 *
 * values in two's complement form    // X, Z, Y
	 * X value: _buff[1] is MSB & _buff[0] is LSB
	 * Y value: _buff[4] is MSB & _buff[5] is LSB
	 * Z value: _buff[2] is MSB & _buff[3] is LSB
 */
void HMC5883L::read()
{
	readFrom(HMC5883L_AxisXDataRegisterMSB, HMC5883L_TO_READ, _buff); //Read HMC5883L_TO_READ bytes from HMC5883L_AxisXDataRegisterMSB onwards and store it in _buff

	magVector[0] = (((int)_buff[0]) << 8) | _buff[1]; // X
	magVector[2] = (((int)_buff[2]) << 8) | _buff[3]; // Z
	magVector[1] = (((int)_buff[4]) << 8) | _buff[5]; // Y

	//magVector[0] *= -1;
	//magVector[1] *= -1;
	//magVector[2] *= -1;
}

/*
 * scale method:
 * apply HMC5883L_SCALE_FACTOR to 'orientationVector' data
 */
void HMC5883L::scaleGain()
{
	magVector[0] *= HMC5883L_SCALE_FACTOR;
	magVector[1] *= HMC5883L_SCALE_FACTOR;
	magVector[2] *= HMC5883L_SCALE_FACTOR;
}

/*
 * applyCalibration method:
 * read max/min range values from EEPROM memory sector and apply offset to 'orientationVector' data
 *
 * range values stored in EEPROM by calibration algorithm
 */
void HMC5883L::applyCalibration()
{
	int EEPROM_ADDRESS=0;  // eeprom offset for magnetometer range values
	float valueRange[6]; //MAG_X_MAX, MAG_X_MIN, MAG_Y_MAX, MAG_Y_MIN, MAG_Z_MAX, MAG_Z_MIN

	for(int i=0;i<6;i++)
	{
		EEPROM.get(EEPROM_ADDRESS + (i*(sizeof(float))),valueRange[i]);
	}

	float offsetX = (valueRange[0] + valueRange[1]) / 2.0f;
	float offsetY = (valueRange[2] + valueRange[3]) / 2.0f;
	float offsetZ = (valueRange[4] + valueRange[5]) / 2.0f;

	//MAGN_X_SCALE = (100.0f / (MAGN_X_MAX - MAGN_X_OFFSET))
	float mag_x_scale = (100.0f / (valueRange[0] - offsetX));
	float mag_y_scale = (100.0f / (valueRange[2] - offsetY));
	float mag_z_scale = (100.0f / (valueRange[4] - offsetZ));

	// magnetom[0] = (magnetom[0] - MAGN_X_OFFSET) * MAGN_X_SCALE;
	magVector[0] = (magVector[0] - offsetX) * mag_x_scale;
	magVector[1] = (magVector[1] - offsetY) * mag_y_scale;
	magVector[2] = (magVector[2] - offsetZ) * mag_z_scale;
}

void HMC5883L::tiltCompensation(float accRoll, float accPitch, float (&data)[3])
{
	float cosPitch = cos(accPitch);
	float sinPitch = sin(accPitch);
	float cosRoll = cos(accRoll);
	float sinRoll = sin(accRoll);

	//Tilt Compensation
	float compMag[3] = {0,0,0};
	compMag[0] = (magVector[0]*cosPitch) + (magVector[2]*sinPitch);
	compMag[1] = (magVector[0]*sinRoll*sinPitch) + (magVector[1]*cosRoll) - (magVector[2]*sinRoll*cosPitch);
	compMag[2] = (-magVector[0]*cosRoll*sinPitch) + (magVector[1]*sinRoll) - (magVector[2]*cosRoll*cosPitch);

	memcpy(data, compMag, sizeof(data));
}

void HMC5883L::setDeclination(float degree, float minutes)
{
	declination = (degree+minutes/60)*(M_PI/180);
}

void HMC5883L::getHeading(float compX, float compY, boolean declinationFlag)
{
		float headingRad;

		//Only needed if the heading value does not increase when the magnetometer is rotated clockwise
		//compY = -compY;

		headingRad = atan2(compY, compX);

		if (declinationFlag == true)
		{
			headingRad += declination;
		}

		//Correct for reversed heading
		if (headingRad < 0)
			headingRad += 2*M_PI;

		//Check for wrap and compensate
		if (headingRad > 2*M_PI)
			headingRad -= 2*M_PI;

		heading = headingRad;
}

/*
 * getOrientationVector method:
 * store calibrated and scaled accelerometer data(x,y,z) in '&data' parameter
 */
void HMC5883L::getMagVector(float (&data)[3])
{
	read();
	applyCalibration();
	scaleGain();

	memcpy(data, magVector, sizeof(data));
}
